File: 02/02.rb
# frozen_string_literal: true

raw_data = ARGV.first || File.read('input')

memory = raw_data.strip.split(',').map(&:to_i)

def execute(memory, noun, verb)
  memory = memory.dup

  memory[1] = noun
  memory[2] = verb

  address = 0

  loop do
    opcode = memory[address]

    break if opcode == 99

    arg1 = memory[memory[address + 1]]
    arg2 = memory[memory[address + 2]]
    result_index = memory[address + 3]

    memory[result_index] =
      case opcode
      when 1 then arg1 + arg2
      when 2 then arg1 * arg2
      else raise "Unknown opcode #{opcode} at address #{address}"
      end

    address += 4
  end

  memory
end

puts 'Part 1: ', execute(memory, 12, 2).map(&:to_s).join(',')

(0..99).each do |noun|
  (0..99).each do |verb|
    if 19690720 == execute(memory, noun, verb).first
      puts "Part 2: noun=#{noun} verb=#{verb} 100 * noun + verb = #{100 * noun + verb}"
    end
  end
end

File: 02/02.go
package main

import (
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
)

func atoi(s string) (result int) {
	result, _ = strconv.Atoi(s)
	return
}

func loadMemory(rawData string) (memory []int) {
	s := strings.Split(string(rawData), ",")
	memory = make([]int, len(s))

	for i := 0; i < len(memory); i++ {
		memory[i] = atoi(s[i])
	}

	return
}

func execute(memory []int, noun int, verb int) []int {
	memory = append([]int(nil), memory...)

	memory[1] = noun
	memory[2] = verb

	address := 0
	for {
		opcode := memory[address]

		if opcode == 99 {
			break
		}

		arg1 := memory[memory[address+1]]
		arg2 := memory[memory[address+2]]
		resultAddress := memory[address+3]
		result := 0

		switch opcode {
		case 1:
			result = arg1 + arg2
		case 2:
			result = arg1 * arg2
		default:
			panic(fmt.Sprintf("Unknown opcode %d at address %d", opcode, address))
		}

		memory[resultAddress] = result

		address += 4
	}

	return memory
}

func main() {
	rawData, err := ioutil.ReadFile("input")
	if err != nil {
		panic(err)
	}

	memory := loadMemory(string(rawData))

	part1 := execute(memory, 12, 2)
	fmt.Printf("Part 1: %d\n", part1[0])

	for x := 0; x <= 99; x++ {
		for y := 0; y <= 99; y++ {
			part2 := execute(memory, x, y)

			if part2[0] == 19690720 {
				fmt.Printf("Part 2: noun=%d verb=%d 100 * noun + verb = %d\n", x, y, 100*x+y)
				return
			}
		}
	}
}
